### PR TITLE
Add production-ready `LangfuseEvaluationDataset` and intent detection evaluation pipeline

### PR DESCRIPTION
**Pull request by ElenaKhaustova**
## Summary
Relates to https://github.com/kedro-org/kedro-plugins/issues/1321

- Productionalized `LangfuseEvaluationDataset` — a Kedro dataset that bridges local JSON/YAML files with remote Langfuse datasets, supporting `local` and `remote` sync policies, versioned snapshots, deduplication, metadata propagation, and comprehensive validation.
- Added `intent_detection_evaluation` pipeline that runs the intent detection agent against a labeled dataset and scores results with two automated evaluators (intent accuracy + LLM-as-a-judge reasoning quality), publishing experiments to Langfuse.
- Added a Jupyter notebook with 13 end-to-end test scenarios covering local mode, remote mode, edge cases, and lifecycle/integration operations.
- Cleaned up the POC evaluation pipeline, refactored configs to use `catalog_*` naming convention, and updated README with evaluation documentation.

## Purpose

The primary goal of this PR is to **assess the functionality of `LangfuseEvaluationDataset`** before moving it to the `kedro-datasets` experimental package. The dataset, evaluation pipeline, and E2E tests together validate that the dataset works correctly across all modes and edge cases in a real Kedro project context.

## What changed

**`LangfuseEvaluationDataset` (productionalized from POC):**
- Input validation with dedicated `_validate_*` static methods
- Exception handling using `LangfuseApiError` / `LangfuseNotFoundError`
- Two sync policies: `local` (local file seeds remote, `save()` merges back) and `remote` (remote-only, supports versioned snapshots via `version` parameter)
- Deduplication by `id` on sync and save, with warnings for id-less items
- Structured logging for all lifecycle events
- Lifecycle operations (update, archive, delete) delegated to the native Langfuse API
- Comprehensive docstrings for mkdocs rendering

**`intent_detection_evaluation` pipeline:**
- Runs the intent detection agent on each evaluation item with isolated LangGraph threads
- Scores with `intent_accuracy` (binary) and `reason_quality` (LLM judge, 1–5)
- Records Langfuse observations linked to prompt version and model
- Experiment naming based on `intent_prompt_version` and `model_name` parameters

**Other changes:**
- Renamed `genai-config.yml` → `catalog_genai_config.yml`, added `catalog_evaluation.yml`
- Simplified config patterns in `settings.py`
- Extended evaluation dataset with borderline test cases (11 items)
- Removed POC evaluation pipeline
- Updated README with evaluation section, fixed typos, updated doc links to `latest`
- Added `langfuse>=3.14.0` to `requirements.txt`

## Design decision: lifecycle delegation

Langfuse supports add, update, delete, and archive operations on dataset items. We chose to **delegate lifecycle management to the native Langfuse API** rather than abstracting these operations into `LangfuseEvaluationDataset`. Rationale:

- **Scope**: The dataset's responsibility is load/save — bridging Kedro pipelines with Langfuse datasets. Lifecycle operations are administrative tasks that happen outside pipeline runs.
- **API surface**: Wrapping every Langfuse operation would duplicate their SDK and create a maintenance burden that grows with each Langfuse release.
- **Flexibility**: Users already have access to the native `Langfuse` client for tracing. Using the same client for lifecycle operations is consistent and doesn't require learning a custom abstraction.

The E2E notebook (Scenario 12) validates this approach: items are updated, archived, and deleted via the native API, then `LangfuseEvaluationDataset.load()` correctly reflects all changes.

### 1. Run the evaluation pipeline

```bash
kedro run -p intent_detection_evaluation --params intent_prompt_version=1,model_name=gpt-4o
```

Check results in the Langfuse UI under Experiments → `intent_prompt_v1_model_gpt-4o`.

### 2. Run the E2E notebook

Open `notebooks/e2e_langfuse_evaluation_dataset.ipynb` and run all cells top-to-bottom. The notebook contains 13 scenarios across 4 sections:

| Section | Scenarios | What it covers |
|---------|-----------|----------------|
| 1. Local mode | 1–4 | Fresh start, idempotent reload, incremental sync, `save()` with dedup |
| 2. Remote mode | 5–7 | Remote load, `save()` without local file, versioned snapshots |
| 3. Edge cases & validation | 8–11 | Items without `id`, metadata passing, config conflicts, invalid credentials |
| 4. Lifecycle & integration | 12–13 | Native API update/archive/delete, running an experiment |

Each scenario has assertions and Langfuse UI verification steps. Verify created datasets are visible in the Langfuse UI after each section.

### 3. Prerequisites

- Langfuse credentials in `conf/local/credentials.yml`
- `langfuse>=3.14.0` installed
- OpenAI API key configured (for the evaluation pipeline's LLM judge)

## Next steps

- Move `LangfuseEvaluationDataset` to the `kedro-datasets` experimental package
- Add unit tests in the `kedro-datasets` repository

### CODE REVIEW

File: kedro-agentic-workflows/data/intent_detection/evaluation/intent_evaluation.json [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+      "reason": "User implies an incident but provides no details about what happened or what they need."
+    }
+  },
+  {
+    "id": "intent_011",
+    "input": {
+      "question": "How long does it usually take to get a claim approved? I've been waiting for weeks."
+    },
+    "expected_output": {
+      "intent": "existing_claim_question",
```
Comment (ravi-kumar-pilla) [inline kedro-agentic-workflows/data/intent_detection/evaluation/intent_evaluation.json:108]: Do we have a fixed set of intent classification strings ?

Comment (ravi-kumar-pilla) [reply kedro-agentic-workflows/data/intent_detection/evaluation/intent_evaluation.json:108]: Oh I see them here - kedro-agentic-workflows/data/intent_detection/prompts/intent_prompt_langfuse.json

File: kedro-agentic-workflows/requirements.txt [hunk 2]
```diff
@@ -4,9 +4,10 @@ ipython>=8.10
 jupyterlab>=3.0
 notebook
 kedro>=1.2.0
-kedro-datasets[pandas-sqltabledataset,pandas-sqlquerydataset,yaml-yamldataset,opik,langchain-chatopenaidataset, langfuse]~=9.0.0
+kedro-datasets[pandas-sqltabledataset,pandas-sqlquerydataset,yaml-yamldataset,opik,langchain-chatopenaidataset,langfuse]~=9.0.0
 kedro-viz>=12.3.0
+langfuse>=3.14.0
 langgraph~=1.0.0
-scikit-learn~=1.7.2
+scikit-learn~=1.7.0
```
Comment (ravi-kumar-pilla) [inline kedro-agentic-workflows/requirements.txt]: is this an intentional downgrade of min version ?

**Comment by ElenaKhaustova** [comment]
## Follow-up changes based on feedback

Based on @ravi-kumar-pilla testing and feedback, the sync/save logic in LangfuseEvaluationDataset was changed from filter-and-skip to upsert-all:

**Problem:** When a user modified an existing item in the local file (e.g. changed the expected_output but kept the same id), the change was silently ignored on the next load() because the old logic only uploaded items whose id didn't already exist on remote.

**Fix:** Both load() (local mode) and save() now upsert all items via Langfuse.create_dataset_item(), which performs a server-side create-or-update by id.

This means:
- Items with an id that already exists on remote are updated in place
- New items are created
- Items without an id always create new entries (with a logged warning)
- Local file merge (save() in local mode) now gives new data precedence over existing entries with the same id

What changed:

- Removed _filter_new_items method entirely
- Updated _sync_local_to_remote and save() to upsert all items
- Updated _merge_items so new items take precedence over existing for same-id entries
- Updated all docstrings, README, and E2E notebook scenarios to reflect upsert semantics

**How `load()` / `save()` now work:**

| | `local` mode | `remote` mode |
|---|---|---|
| **`load()`** | Upserts all local items to remote, returns `DatasetClient` | Fetches remote as-is (supports `version` for snapshots) |
| **`save(data)`** | Upserts all items to remote + merges into local file (new data takes precedence) | Upserts all items to remote, no local file interaction |

Items with an `id` are created or updated; items without `id` always create new entries (warned).

**Comment by ravi-kumar-pilla** [comment]
Hi @ElenaKhaustova , The sync mode is easy to understand now, thank you. 

I have a question for `remote` mode `load()` operation. This works as mentioned in the matrix. 

Since we say it as sync mode, does it make sense to update the local file if the content is different from remote when someone has a remote mode and loads the dataset ? Similar to how we upsert remote on `local` mode `load()` operation ?

**Comment by ravi-kumar-pilla** [comment]
Tested both the functionalities via CLI and notebook. They work great. For the scenario 13 in notebook, though the experiment is getting created, I get the below error. Again something related to a http error but rest everything works as expected.

image

Thank you

**Comment by ElenaKhaustova** [comment]
> Hi @ElenaKhaustova , The sync mode is easy to understand now, thank you.
> 
> I have a question for `remote` mode `load()` operation. This works as mentioned in the matrix.
> 
> Since we say it as sync mode, does it make sense to update the local file if the content is different from remote when someone has a remote mode and loads the dataset ? Similar to how we upsert remote on `local` mode `load()` operation ?

Good question — I considered this but decided against it. Remote mode's contract is explicitly "no local file interaction," and changing that may introduce surprising side effects:

- **Data loss risk**: if someone has a curated local file and switches to `remote`, the remote content would overwrite their work
- **Format mismatch**: remote items have Langfuse-generated fields (timestamps, auto-IDs) that may don't map cleanly to the simple local authoring format
- **Rremote users may not configure `filepath` at all** — it's optional in remote mode

The two modes serve intentionally different workflows: `local` = local file is the authoring surface that pushes to remote; `remote` = remote is the sole authority, local file is irrelevant.

On the other hand, we've done it differently for **`LangfusePromptDataset`**. But the key here is that it manages a single prompt — one small, self-contained piece of text. Writing it back to a local file is essentially caching a 1:1 copy. The local file is always an exact mirror of the remote prompt, so there's no ambiguity about what gets written.

**`LangfuseEvaluationDataset`** manages a **collection of items** that can diverge significantly between local and remote:
- Remote may have items added via native API that were never in the local file
- Remote items carry server-generated fields (timestamps, auto-IDs)
- Some items may be archived or deleted on remote
- The local file is an **authoring surface**, not a cache — writing the full remote state back would mix server-generated data into the user's curated file
- Datasets can be much larger than prompts, use may not want to keep the copy locally.
- `filepath` is **optional** — most remote-mode users won't set it at all

So the precedent exists, but the data shape is fundamentally different: a single prompt vs. an evolving collection with independent lifecycle operations. If consistency across datasets becomes a priority, we can add a `mirror` sync policy later to opt into `remote` to `local` sync explicitly.

TLDR: I think we should probably proceed with the simpler option and revisit it later if there's a concrete use case.

**Comment by ElenaKhaustova** [comment]
@ravi-kumar-pilla 

> Again something related to a http error but rest everything works as expected.

Did you use the right host?

```bash
LANGFUSE_BASE_URL = "https://cloud.langfuse.com" # 🇪🇺 EU region
LANGFUSE_BASE_URL = "https://us.cloud.langfuse.com" # 🇺🇸 US region
```

**Comment by ravi-kumar-pilla** [comment]
> TLDR: I think we should probably proceed with the simpler option and revisit it later if there's a concrete use case.

Sounds good. May be we mention this in the documentation. Thank you

**Review by lrcouto — APPROVED** [review]
